compiler/src/dotty/tools/dotc/interpreter/Interpreter.scala

```diff
@@ -0,0 +1,129 @@
+   */
+  def interpretTree[T](tree: Tree)(implicit ct: ClassTag[T]): Option[T] = {
+    try {
+      interpretTreeImpl(tree) match {
+        case obj: T => Some(obj)
+        case obj =>
+          ctx.error(s"Interpreted tree returned a result of an unexpected type. Expected ${ct.runtimeClass} but was ${obj.getClass}", tree.pos)
+          throw new StopInterpretation
+      }
+    } catch {
+      case _: StopInterpretation => None
+    }
+  }
+
+  /** Returns the interpreted result of interpreting the code represented by the tree.
+   *  Returns the result of the interpreted tree.
+   *
+   *  If some error is encountered while interpreting a ctx.error is emited and a StopInterpretation is thrown.
+   */
+  private def interpretTreeImpl(tree: Tree): Object = {
+    try {
+      tree match {
+        case Apply(_, quote :: Nil) if tree.symbol eq defn.quoteMethod =>
+          new RawExpr(quote)
+        case TypeApply(_, quote :: Nil) if tree.symbol eq defn.typeQuoteMethod =>
+          new RawType(quote)
+
+        case Literal(Constant(c)) =>
+          c.asInstanceOf[AnyRef]
+
+        case Apply(fn, args) if fn.symbol.isConstructor =>
+          val cls = fn.symbol.owner
+          val clazz = classLoader.loadClass(cls.symbol.fullName.toString)
+          val paramClasses = paramsSig(fn.symbol)
+          val args1: List[Object] = args.map(arg => interpretTreeImpl(arg))
+          clazz.getConstructor(paramClasses: _*).newInstance(args1: _*).asInstanceOf[Object]
+
+        case Apply(fun, args) if fun.symbol.isStatic =>
+          val clazz = classLoader.loadClass(fun.symbol.owner.companionModule.fullName.toString)
+          val paramClasses = paramsSig(fun.symbol)
+          val args1: List[Object] = args.map(arg => interpretTreeImpl(arg))
+          val method = clazz.getMethod(fun.symbol.name.toString, paramClasses: _*)
+          method.invoke(null, args1: _*)
+
+        case tree: RefTree if tree.symbol.isStatic =>
+          val clazz = classLoader.loadClass(tree.symbol.owner.companionModule.fullName.toString)
+          val method = clazz.getMethod(tree.name.toString)
+          method.invoke(null)
+
+        case tree: RefTree if tree.symbol.is(Module) =>
+          ??? // TODO
+
+        case Inlined(_, bindings, expansion) =>
+          if (bindings.nonEmpty) ??? // TODO evaluate bindings and add environment
+          interpretTreeImpl(expansion)
+        case _ =>
+          val msg =
+            if (tree.tpe.derivesFrom(defn.QuotedExprClass)) "Quote needs to be explicit or a call to a static method"
+            else "Value needs to be a explicit or a call to a static method"
+          ctx.error(msg, tree.pos)
+          throw new StopInterpretation
+      }
+    } catch {
+      case ex: NoSuchMethodException =>
+        ctx.error("Could not find interpreted method in classpath: " + ex.getMessage, tree.pos)
+        throw new StopInterpretation
```

Comment:
Seems like there's repetition worth factoring out.

Reply:
But having two try/catches in `interpretTree` and `interpretTreeImpl` looks like overkill. Maybe propagate everything to `interpretTree`?

Reply:
The issue is that the error message should know which method or class was not found which is not accessible in `interpretTree`. Instead I would create local methods for `loadClass` and `invoke` that would catch the Java exceptions and throw the `StopInterpretation` with the error message and position.

Reply:
OK, makes sense.